sycl/source/detail/platform_impl.cpp

```diff
@@ -120,96 +121,147 @@ vector_class<platform> platform_impl::get_platforms() {
-
-    // Skip opening sequence "{{"
-    str += 2;
-
-    *valuePtr = str;
+    else if ((AllowList.compare(Pos, DriverVersion.size(), DriverVersion)) ==
+             0) {
+      DecDescs.back().DevDriverVer =
+          getValue(AllowList, Pos, DriverVersion.size());
+      if (AllowList[Pos] == ',') {
+        Pos++;
+      }
+    }
 
-    // Increment until closing sequence is encountered
-    while (('\0' != *str) && ('}' != *str || '}' != *(str + 1)))
-      ++str;
+    else if ((AllowList.compare(Pos, PlatformName.size(), PlatformName)) == 0) {
+      DecDescs.back().PlatName = getValue(AllowList, Pos, PlatformName.size());
+      if (AllowList[Pos] == ',') {
+        Pos++;
+      }
+    }
 
-    if ('\0' == *str)
-      throw sycl::runtime_error("Malformed device allowlist", PI_INVALID_VALUE);
+    else if ((AllowList.compare(Pos, PlatformVersion.size(),
+                                PlatformVersion)) == 0) {
+      DecDescs.back().PlatVer =
+          getValue(AllowList, Pos, PlatformVersion.size());
+    } else if (AllowList.find('|', Pos) != std::string::npos) {
+      Pos = AllowList.find('|') + 1;
+      while (AllowList[Pos] == ' ') {
+        Pos++;
+      }
+      DecDescs.emplace_back();
+    }
 
-    *size = str - *valuePtr;
+    else {
+      throw sycl::runtime_error("Unrecognized key in device allowlist",
+                                PI_INVALID_VALUE);
+    }
+  } // while (Pos <= AllowList.size())
+  return DecDescs;
+}
 
-    // Skip closing sequence "}}"
-    str += 2;
+std::vector<int> convertVersionString(std::string Version) {
+  // version string format is xx.yy.zzzzz.ww   WW is optional
+  std::vector<int> Values;
+  size_t Pos = 0;
+  size_t Start = Pos;
+  if ((Pos = Version.find(".", Pos)) == std::string::npos) {
+    throw sycl::runtime_error("Malformed syntax in version string",
+                              PI_INVALID_VALUE);
+  }
+  Values.push_back(std::stoi(Version.substr(Start, Pos - Start)));
+  Pos++;
+  Start = Pos;
+  if ((Pos = Version.find(".", Pos)) == std::string::npos) {
+    throw sycl::runtime_error("Malformed syntax in version string",
+                              PI_INVALID_VALUE);
+  }
+  Values.push_back(std::stoi(Version.substr(Start, Pos - Start)));
+  Pos++;
+  size_t Prev = Pos;
+  if ((Pos = Version.find(".", Pos)) == std::string::npos) {
+    Values.push_back(std::stoi(Version.substr(Prev)));
+  } else {
+    Values.push_back(std::stoi(Version.substr(Start, Pos - Start)));
+    Pos++;
+    Values.push_back(std::stoi(Version.substr(Pos)));
+  }
+  return Values;
+}
 
-    if ('\0' == *str)
-      break;
+enum MatchState { UNKNOWN, MATCH, NOMATCH };
 
-    // '|' means that the is another filter
-    if ('|' == *str)
-      decDescs.emplace_back();
-    else if (',' != *str)
-      throw sycl::runtime_error("Malformed device allowlist", PI_INVALID_VALUE);
+MatchState matchVersions(std::string Version1, std::string Version2) {
```

Comment:
NIT: does it make sense to move matching-related code to a separate source, to avoid clutter?

Reply:
It does not look like this in the source.  The diff makes it looks confusing.  Between the pink and the green, getAllowListDesc() ends in the normal way, and there is the convertVersionString() function .  Then we find matchVersions().